src/ast/ast_validations.c

```diff
@@ -914,26 +950,62 @@ static AST_Validation _Validate_RETURN_Clause(const AST *ast) {
 }
 
 static AST_Validation _Validate_UNWIND_Clauses(const AST *ast) {
-	const cypher_astnode_t **unwind_clauses = AST_GetClauses(ast, CYPHER_AST_UNWIND);
-	if(!unwind_clauses) return AST_VALID;
+	AST_Validation res               =  AST_VALID;
+	uint           *clause_indices   =  AST_GetClauseIndices(ast, CYPHER_AST_UNWIND);
+	uint           clause_count      =  array_len(clause_indices);
+	rax            *defined_aliases  =  NULL;
+
+	if(clause_count == 0) goto cleanup;
 
-	AST_Validation res = AST_VALID;
-	uint clause_count = array_len(unwind_clauses);
 	for(uint i = 0; i < clause_count; i++) {
-		const cypher_astnode_t *expression = cypher_ast_unwind_get_expression(unwind_clauses[i]);
-		// Verify that all elements of the UNWIND collection are supported by RedisGraph
+		uint clause_idx = clause_indices[i];
+		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root,
+																	 clause_idx);
+		const cypher_astnode_t *expression = cypher_ast_unwind_get_expression(clause);
+		// verify that all elements of the UNWIND collection
+		// are supported by RedisGraph
 		uint child_count = cypher_astnode_nchildren(expression);
 		for(uint j = 0; j < child_count; j ++) {
 			res = CypherWhitelist_ValidateQuery(cypher_astnode_get_child(expression, j));
 			if(res != AST_VALID) goto cleanup;
 		}
-		// Verify that UNWIND doesn't call non-existent or unsupported functions.
+		// verify that UNWIND doesn't call non-existent or unsupported functions
 		res = _ValidateFunctionCalls(expression, true);
 		if(res != AST_VALID) goto cleanup;
 	}
 
+	uint start_offset = 0;
+	for(uint i = 0; i < clause_count; i ++) {
+		uint clause_idx = clause_indices[i];
+		defined_aliases = raxNew();
+
+		// collect all entities that are bound before this UNWIND clause
+		for(uint j = start_offset; j < clause_idx; j ++) {
+			const cypher_astnode_t *prev_clause = cypher_ast_query_get_clause(
+													  ast->root, j);
+			_AST_GetDefinedIdentifiers(prev_clause, defined_aliases);
+		}
```

Comment:
Because `defined_aliases` is freed at each iteration of the outer loop I think we might miss cases such as:
`UNWIND [1] AS a UNWIND [2] AS b UNWIND [3] AS c UNWIND [4] AS a`

Seems to me like there's no need to free `defined_aliases` within the inner loop.

Reply:
Good call, fixed!